crypto/getenv.c

```diff
@@ -15,6 +15,50 @@
 #include "internal/cryptlib.h"
 #include "internal/e_os.h"
 
+/*
+ * Additional checks for securely allowing access to selected environment variables
+ */
+#if defined (OPENSSL_NETCAP_ALLOW_ENV) && defined(__linux__)
+
+/*
+ * Utility function to check if a string is in an array of strings. returns true if found
+ */
+int str_in_list(const char *candidate, const char *list[], int n)
+{
+  if ( (candidate!=NULL) && (list!=NULL))
+  {
+    int len = n;
+    if (len>=1)
+    {
+        int i;
+        for ( i=0; i<len; i++)
+        {
+            if (strcmp(candidate,list[i])==0)
+              // found
+              return 1;
+        }
+    }
+  }
+  // drop-through case -- not found (or invalid parameters)
+  return 0;
+}
+
+char *restricted_getenv(const char *name)
+{
+    /* Environment variables to permit in secure mode (subject to conditions). Extend as needed */
+    const char * permitted_env [ ] = { "OPENSSL_MODULES", "OPENSSL_CONF" };
```

Comment:
This feels a bit like we're hardcoding policy to me (which may or may not be ok).  Currently, ossl_safe_getenv is called to fetch the following environment variables:

OPENSSL_CONF_INCLUDE
CTLOG_FILE_EVP
OPENSSL_CONF
OPENSSL_ENGINES
no_proxy
OPENSSL_NO_PROXY
OPENSSL_HTTP_PROXY
OPENSSL_HTTPS_PROXY
OPENSSL_MODULES
RANDFILE
HOME
X509_CERT_DIR_EVP

Seeing this makes me ask "which of those variables do we want to allow to be fetched when in a secure environment, which do we want to hide, and under what conditions?"

The more I look at it, the more I think the only reasonable answer is "we don't know" or more specifically "we can't know".  I can see how this works in your situation because you specifically need those environment variables, but I could just as easily see another user specifically wanting to restrict access to them when in a secure environment, or they may need a superset of them, some of which your use case may consider to be a security risk.  As such, it seems wrong to hardcode them.

A better solution might be to create an OPENSSL_ALLOW_ENV variable that can be read by ossl_safe_getenv which holds a list of env vars that should be fetched using getenv rather than secure_getenv.  That would allow users to set a selected list of trusted variables at run time to be passed through from outside the secure environment.

Of course, such an implementation would require that OPENSSL_ALLOW_ENV either be retrieved via getenv (which negates the security introduced by secure_getenv), or require that it be set within the secure environment, which puts you back to needing a wrapper script.

Reply:
This is a critical design question. 

The intent of this draft was to demonstrate a mechanism, but also point out the exact list needs discussion, and could be configuration dependent.

It cannot be an environment variable (at least with the current design) since that effectively would allow all values. We have a specific issue with the configuration needed for quantum-safe encryption, therefore I decided to place a hardcoded set initially. I agree that ultimately a hard-coded set is not desirable.

Note that these extra variables should not be read in all secure environments, only if the ONLY capability set is to allow ports <1024 to be allocated. In other cases they should be hidden.

The list could be defined in openssl.cnf. However if the default did not permit these values we'd be back to needing deployment-time configuration.?

Would we be able to get consensus on a default set?

Reply:
Thats the thing I'm strugging with here, I don't think we can agree on a default set, because eventually, someone will say "Can we add the variable X to the list?".  And perhaps we can, or decide we can't, but its going to be a game of whack-a-mole one way or ther other, explaining to subsequent users why we think their variable can or can't be on the list.  The only way I see to avoid that, is to let users decide whats safe in their environment, for which we either need an allowlist environment variable (with the recursive inside/outside the secure boundary issues that brings with it) or a command line --allow-env option on any application that wants to make use of this feature.

The capabilies set is the same problem in my mind. restricting this feature to only work when CAP_NET_BIND_SERVICE is the only capability set makes sense for this use case, but there are lots of capabilities that I could envision users wanting to set that would also necessitate the use of this mechanism.  It makes more sense in my mind to use a mechanism like this always (i.e. assume we are in a secure environment, and if secure_getenv fails, back off to using getenv, but only for those variables which a user has informed us at run time are considered safe to fetch from outside the secure environment

Reply:
Seems the options available to specify which are exempt are:

* API call
* Command line argument
* Configuration
* Environment variable (although the consensus seems to be that this is a bad idea)
* Hard coded list (too difficult to define globally)

Reply:
To be complete, there is another option: do nothing.  Its still entirely possible to wrap an application in a capabilities adjusted shell wrapper that explicitly sets any needed environment variables and execs the needed openssl linked application. Doing so gives the user the most control over what environment variables are allowed in and under what conditions they feel comfortable with

Reply:
Why did we think this approach could be useful?

There are many services listening on standard (<1024) ports which make use of OpenSSL. This might include apps like postgres, nodejs, hginx, haproxy, openssh, a wide range of https applications. Some of these apps themselves are used in front of others (on >1024) as proxies/routers

Each of these will differ in terms of the process model - which would affect how wrappers could be introduced.

Each are likely working on specific plans to use new encryption algorithms, but this will take time.

Having a common approach across all these tools -- for this specific case (there is no issue > 1024) -- would allow quicker adoption of quantum-safe encryption algorithms, with likely no app change, by appropriately configuring those algorithms & making the executable modules available through the environment variables listed - as can already be done for apps on ports >1024 thanks to the provider support in 3.x, and implementations such as liboqs. 
Since my understanding is that the net bind capability is a very specific capability, that this would be ok. 

Wrappers, authbind, application-specific coding, system-wide configurations remain other options on a by application/deployment basis

Reply:
I agree this is useful, I have no  argument there.

My concern is for whom else this might be useful (or more specifically useful if we just expand the capabilities for which it can be used).

To cite your example of applications which set CAP_NET_BIND_SERVICE, looking at nginx, it also has code that sets CAP_NET_RAW when NGX_HAVE_CAPABILITIES is defined in that build.  In such a build, this change becomes "non-useful", in that OPENSSL_* variables become hidden again, when they otherwise might want them.

So, the question: Do we add CAP_NET_RAW to the filter list, effectively saying only those two are allowed?  Or do we say no, only CAP_NET_BIND_SERVICE, because we think thats safe?

My argument is that we shouldn't be the gatekeepers of what we think is safe in this case.  If we're going to use secure_getenv, we need to use it, as it protects us from malicious intrusion via injected environment variables.

If a user of openssl thinks something is safe, thats their business, but they need work with the environment that the os is giving us.  I understand that will be different for different deployments, but thats the only way to place this policy in the hands of the users making that decision.

If we want to help them manage that, I think thats fine and good, but we shouldn't code that policy into the library itself.  I think the best we can do is either:
a) guide users in how to setup the secure environment properly (via a wrapper script or some such)
or
b) give them some tools to implement such a policy on their own (i.e. adding a command line switch to apps we distribute to dynamically create an alowlist of variables they consider safe to read from outside the secure environment at their own risk)

Reply:
@nhorman Good questions, and that's an interesting observation on NGINX. I agree configuration needs to be in the control of the user (the person building the code for now).

I'm just delaying further updates briefly which I go back to check out use cases, particularly in the context of wrappers (and/or authbind) before looking at the remaining comments, and in particular a proposal around configurability